### ISSUE TITLE
Feature request - use_package call

### ISSUE THREAD

**Issue by rsangole** [KentonWhite/ProjectTemplate#258]
#### Report an Issue / Request a Feature

I'm submitting a (Check one with "x") : 
- [ ] bug report 
- [x] feature request

***

##### Issue Severity Classification - 

(Check one with "x") : 
- [ ] 1 - Severe
- [ ] 2 - Moderate
- [x] 3 - Low

##### Expected Behavior

An easy console driven method to quickly add libraries to `global.dcf`, which would utilize autocomplete for the package names.

##### Current Behavior

Today, as we work on the projects, adding a library to the `global.dcf` file is a manual process. When many `.R` files are open, and the open folder is different, it's a cumbersome workflow to open `global.dcf` and add the library name without any autocomplete option. (Results in spelling mistakes like I did today - `rtsne` instead of `Rtsne`.

##### Possible Solution

Something along the lines of the devtools::use_package()` which can be called from command line as well as will autocomplete package names while typing.

**Comment by KentonWhite** [comment]
This would be very handy!  Please keep in mind the CRAN policy:

> Packages should not write in the user’s home filespace (including clipboards), nor anywhere else on the file system apart from the R session’s temporary directory (or during installation in the location pointed to by TMPDIR: and such usage should be cleaned up). Installing into the system’s R installation (e.g., scripts to its bin directory) is not allowed.

> Limited exceptions may be allowed in interactive sessions if the package obtains confirmation from the user.

**Comment by rsangole** [comment]
@KentonWhite it's interesting they have this policy when tools like `devtools` write to a file in the user's home filespace, and tools like `clipr` write to the user's clipboard.

**Comment by Hugovdberg** [comment]
I think the user issuing this command with the sole intent of writing to the system would fall under the exceptions. It's not like it's a possibly unexpected side effect. 
This feature needs thorough testing though, we want to very sure we don't loose any settings made by the user. We might even mark it as experimental in the documentation so people will use it with caution. If we don't receive any issues we can remove the message later.

**Comment by KentonWhite** [comment]
I agree.  The simplest way to be compliant is ask the user for confirmation before making the change :)

**Comment by rsangole** [comment]
Pseudocode:

- User calls `add_package()`
- Check if package exists in `global.dcf`
  - If yes, inform user. Take no action. Done.
  - If no, ask user for consent to modify `global.dcf`
    - If yes, write to file
    - Optional if `verbose = TRUE`, print new `global.dcf` to console for visual confirmation of change

**Comment by rsangole** [comment]
To extend this logic and help ease workflow, I can also think of other `global.dcf` modifiers along the lines of `munge_on()`, `munge_off()`, `loaddata_on()`, `loaddata_off()`, `loadcache_on()`, `loadcache_off()` etc. You get the idea.

**Comment by rsangole** [comment]
@KentonWhite I wanted to make separate pipable verbs which will work well in the `tidyverse` work flow sense. These will be more human readable and quicker to type in the console. For example:

```
# I make some changes in the underlying input data stored in `data/` . I need to quickly reload data, perform all my munging activities. I can quickly type:

data_on() %>%
   munge_on() %>%
   cache_off() %>%
   reload_project()
```

```
# I add some libraries to my config file. I want to reload all my libraries to ensure they're in my working environment.

add_package(lattice) %>%
   add_pacakge(TSClust) %>%
   reload_libraries()
```

```
# I did something odd. I need to start a fresh from my known-good cache files

munge_off() %>%
   cache_on() %>%
   reload_project()
```

**Comment by rsangole** [comment]
In my 2nd example:
```
loaddata_on() %>%
   munge_on() %>%
   loadcache_off() %>%
   reload_project() %>%
   loaddata_off() %>% munge_off() %>% loadcache_on()
```
I want to be able to take actions on the `dcf` file _after_ running `reload_project()`. Thus, I envisioned this type of piping to enable lazy evaluation. The piping is to build a sequence of actions desired; each pipe will enable a build up for the sequence. At the very end of the pipe, the sequence will be executed. 

I also think the pipe makes code much more readable than operators, but that's just my personal preference.

I agree with the statements about the temporary change (for 1 runtime) vs permanent change. Summarizing our discussion so far, we could have the following:

* `add_package()` to add a new package to `dcf`
* `load.project(...)` and `reload.project(...)` have auto-complete arguments to change popular configs temporarily for 1 execution.
* create a `update.config(..., config = 'global.dcf')` to allow change in config file from the console

**Comment by Hugovdberg** [comment]
But how would the last function in the chain know it is the last function? Put differently, if every function passes the full config, how would each function know whether to write the config to disk or not? The pipe operator is just a nice syntactic sugar for this:

```R
loadcache_on(
  munge_off(
    loaddata_off(
      reload_project(
        loadcache_off(
          munge_on(
            loaddata_on()
          )
        )
      )
    )
  )
)
```
There is nothing magical going on to pass knowledge about functions inside or outside the current function.

I think a better approach for your that example would be like this:

```R
# Set config permanently to use cache only
update.config(
    data.loading = FALSE,
    munging = FALSE,
    cache.loading = TRUE
)
# Reload once from data
reload.project(
    data.loading = TRUE, 
    munging = TRUE, 
    cache.loading = FALSE
)
```

**Comment by Hugovdberg** [comment]
It's not to say I don't like the idea, but the added value of the piping construction is not clear to me. But if you have an elegant solution then I'm open for it :-) I realise I can be a bit stubborn and strongly opinionated at times ;-)

**Comment by Hugovdberg** [comment]
Also, we should add a `template` argument to the `update.config` function, which defaults to `NULL` for the current project, so you can edit the config in one of your templates so all future projects from that template use the new settings.

**Comment by rsangole** [comment]
@Hugovdberg , I always up for a debate over ideas; if my ideas can pass the muster of smarter folks like you two, then I know it's a good idea. Keep it coming!

> how would each function know whether to write the config to disk or not?

I don't know, tbh. In my head, things are still at a conceptual level - I haven't put thought into the actual code structure yet. While I truly (deeply) love pipes, I can agree that your proposal of:

```
# Set config permanently to use cache only
update.config(
    data.loading = FALSE,
    munging = FALSE,
    cache.loading = TRUE,
    template = NULL,
    ...
)
# Reload once from data
reload.project(
    data.loading = TRUE, 
    munging = TRUE, 
    cache.loading = FALSE,
    ...
)
```
achieves the intent of my needs. Let's go with this, along with an`add.package()`. (Instead of `add_package()`, since we're using dots in this package.)
